File: python/src/trezorlib/root_packet.py
# This file is part of the Trezor project.
#
# Copyright (C) SatoshiLabs and contributors
#
# This library is free software: you can redistribute it and/or modify
# it under the terms of the GNU Lesser General Public License version 3
# as published by the Free Software Foundation.
#
# This library is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU Lesser General Public License for more details.
#
# You should have received a copy of the License along with this library.
# If not, see <https://www.gnu.org/licenses/lgpl-3.0.html>.

import logging
import typing as t
from hashlib import sha256

import construct as c
from construct_classes import subcon

from .construct_helpers import Reserved
from .firmware.sanity_struct import SanityCheckedStruct

LOG = logging.getLogger(__name__)
TREZOR_ROOT_PACKET = b"TRRP"
MLDSA_SIG = c.Bytes(2420)

# Number of signature slots in the unauthenticated part. `sigmask` must select
# exactly this many keys.
NUM_SIGNATURES = 2


class RootPacketAuth(SanityCheckedStruct):
    """Authenticated part of the root packet."""

    NAME: t.ClassVar[str] = "root packet (authenticated part)"

    MAX_RINGS: t.ClassVar[int] = 3

    ring_mask: int
    sigmask: int
    timestamp: int
    root_rings: list[bytes]

    SUBCON = c.Struct(
        "_magic" / c.Const(TREZOR_ROOT_PACKET, c.Bytes(4)),
        "_version" / c.Const(b"\x01", c.Bytes(1)),
        "ring_mask" / c.Byte,
        "sigmask" / c.Byte,
        "_reserved" / Reserved(1),
        "timestamp" / c.Int32ul,
        "root_rings" / c.Array(lambda ctx: bin(ctx.ring_mask).count("1"), c.Bytes(32)),
    )

    # ---- higher-level, flag-like abstraction ----

    def has_ring(self, index: int) -> bool:
        """Whether ring `index` is present in the mask."""
        return bool(self.ring_mask & (1 << index))

    @property
    def ring_indices(self) -> list[int]:
        """Ring indices present in the mask, ascending (0..MAX_RINGS-1)."""
        return [i for i in range(self.MAX_RINGS) if self.has_ring(i)]

    def ring(self, index: int) -> bytes:
        """The 32-byte root for ring `index`."""
        if not self.has_ring(index):
            raise KeyError(f"ring {index} not present in mask {self.ring_mask:#04x}")
        # array is stored ascending by ring index; position = set bits below `index`
        pos = bin(self.ring_mask & ((1 << index) - 1)).count("1")
        return self.root_rings[pos]

    @property
    def rings(self) -> dict[int, bytes]:
        """Mapping {ring_index: 32-byte root} for all present rings."""
        return {i: self.ring(i) for i in self.ring_indices}

    def digest(self) -> bytes:
        """SHA256 over the authenticated part.

        This is the value that gets signed.
        """
        return sha256(self.build()).digest()

    def sanity_check(self, image: bytes, errors: t.Sequence[str] = ()) -> None:
        _errors: list[str] = list(errors)

        # `root_rings` length must equal the number of set mask bits
        expected = bin(self.ring_mask).count("1")
        if len(self.root_rings) != expected:
            _errors.append(
                f"root_rings length {len(self.root_rings)} does not match "
                f"mask popcount {expected} (mask {self.ring_mask:#04x})"
            )

        # no bits set above MAX_RINGS
        if self.ring_mask >> self.MAX_RINGS:
            _errors.append(
                f"ring_mask {self.ring_mask:#04x} has bits set above "
                f"MAX_RINGS={self.MAX_RINGS}"
            )

        super().sanity_check(image, _errors)


class RootPacket(SanityCheckedStruct):
    NAME: t.ClassVar[str] = "root packet"

    MAX_RINGS: t.ClassVar[int] = RootPacketAuth.MAX_RINGS

    # Authenticated part -- the digest is computed over this.
    auth: RootPacketAuth = subcon(RootPacketAuth)
    # Unauthenticated part -- signatures of `auth.digest()` only.
    signature_0: bytes
    signature_1: bytes

    SUBCON = c.Struct(
        "auth" / RootPacketAuth.SUBCON,
        "signature_0" / MLDSA_SIG,
        "signature_1" / MLDSA_SIG,
    )

    def digest(self) -> bytes:
        """SHA256 over the authenticated part; see `RootPacketAuth.digest`."""
        return self.auth.digest()

    def sanity_check(self, image: bytes, errors: t.Sequence[str] = ()) -> None:
        _errors: list[str] = list(errors)

        # `sigmask` must select exactly as many keys as there are signatures
        selected = bin(self.auth.sigmask).count("1")
        if selected != NUM_SIGNATURES:
            _errors.append(
                f"sigmask {self.auth.sigmask:#04x} selects {selected} keys, "
                f"expected {NUM_SIGNATURES}"
            )

        super().sanity_check(image, _errors)

File: core/embed/io/app_arena/root_packet.c
/*
 * This file is part of the Trezor project, https://trezor.io/
 *
 * Copyright (c) SatoshiLabs
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

#ifdef KERNEL_MODE

#include <trezor_rtl.h>

#include <io/app_root.h>

#include <sha2.h>

#include "root_packet.h"

static const mldsa44_public_key_t * const ROOT_PACKET_KEYS[] = {
#if defined(BOOTLOADER_DEVEL) || defined(TREZOR_EMULATOR)
    (const mldsa44_public_key_t*) "\xec\x01\xe6\x02\x63\x02\x4f\x7e\x71\x72\x80\x13\xb7\x31\xf7\xba\x12\x99\xf5\x18\xc2\x7b\xa3\xed\x8f\x4a\x21\x99\x74\x12\x7c\x62",
    (const mldsa44_public_key_t*) "\x8a\xf8\x87\x80\x85\x94\x6e\xd8\xb1\x16\xbd\x24\xc0\xf2\xaa\xc4\x8b\x7e\x8f\x11\xbf\x06\x87\x25\xcc\xfb\xb1\x52\xab\xf7\xa4\xcd",
#else
    MODEL_ROOT_PACKET_KEYS
#endif
};

ts_t root_packet_verify(const void* data, size_t size,
                        root_packet_auth_t** out) {
  TSH_DECLARE;
  ts_t status;

  TSH_CHECK_ARG(data != NULL);
  TSH_CHECK_ARG(out != NULL);

  *out = NULL;

  TSH_CHECK(size >= sizeof(root_packet_auth_t), TS_EBADMSG);

  root_packet_auth_t* auth = (root_packet_auth_t*)data;

  TSH_CHECK(auth->magic == ROOT_PACKET_MAGIC, TS_EBADMSG);
  TSH_CHECK(auth->version == ROOT_PACKET_VERSION, TS_EBADMSG);
  TSH_CHECK(auth->ring_mask != 0, TS_EBADMSG);
  TSH_CHECK(auth->ring_mask <= (1 << APP_RING_COUNT) - 1, TS_EBADMSG);
  TSH_CHECK(auth->timestamp != 0, TS_EBADMSG);

  // Calculate the expected size of the authenticated part of the root packet
  size_t auth_part_size =
      sizeof(root_packet_auth_t) +
      sizeof(sha256_digest_t) * __builtin_popcount(auth->ring_mask);

  TSH_CHECK(size == auth_part_size + sizeof(root_packet_unauth_t), TS_EBADMSG);

  // Calculate hash of authenticated part of the root packet
  sha256_digest_t auth_hash;
  SHA256_CTX ctx;
  sha256_Init(&ctx);
  sha256_Update(&ctx, (const uint8_t*)auth, auth_part_size);
  sha256_Final(&ctx, auth_hash.bytes);

  // Verify signatures
  root_packet_unauth_t* unauth =
      (root_packet_unauth_t*)((uint8_t*)auth + auth_part_size);

  uint8_t sigmask = auth->sigmask;
  uint8_t sigmask_inv = 0;  // FIH

  TSH_CHECK(__builtin_popcount(sigmask) == ARRAY_LENGTH(unauth->signature),
            TS_EBADMSG);

  for (int sig_idx = 0; sig_idx < ARRAY_LENGTH(unauth->signature); sig_idx++) {
    // Get the index of the public key in the signature mask
    int key_idx = __builtin_ctz(sigmask);
    TSH_CHECK(key_idx < ARRAY_LENGTH(ROOT_PACKET_KEYS), TS_EBADMSG);

    secbool valid = secfalse;
    status =
        mldsa44_verify(&unauth->signature[sig_idx], &auth_hash,
                       sizeof(auth_hash), ROOT_PACKET_KEYS[key_idx], &valid);
    TSH_CHECK_OK(status);
    TSH_CHECK(valid == sectrue, TS_EBADMSG);

    // Mark the key as used
    sigmask &= ~(1 << key_idx);
    sigmask_inv |= (1 << key_idx);
  }

  // Check that all signatures were verified
  TSH_CHECK(sigmask == 0, TS_EBADMSG);
  TSH_CHECK(sigmask_inv == auth->sigmask, TS_EBADMSG);  // FIH

  *out = auth;

cleanup:
  TSH_RETURN;
}

#endif  // KERNEL_MODE
